packages/@lwc/engine-core/src/framework/hydration.ts

```diff
@@ -755,3 +774,19 @@ function areCompatibleNodes(client: Node, ssr: Node, vnode: VNode, renderer: Ren
 
     return isCompatibleElements;
 }
+
+function haveCompatibleStaticParts(vnode: VStatic, renderer: RendererAPI) {
+    const { owner, parts } = vnode;
+
+    if (isUndefined(parts)) {
+        return true;
+    }
+
+    for (const part of parts) {
+        const hasMatchingAttrs = validateAttrs(part, part.elm!, owner, renderer, () => true);
+        if (isFalse(hasMatchingAttrs)) {
+            return false;
+        }
+    }
+    return true;
+}
```

Comment:
I think an important invariant here is:

1. It's never the case that `parts` is undefined on the server but defined on the client (or vice-versa)
2. It's never the case that `parts` has one length on the server but another on the client

This might be worth a comment.

Reply:
Added a [comment](https://github.com/salesforce/lwc/pull/4056/commits/93f51d8c0d62013ab61fb6f499ce79329df14da1#diff-a9673af211cdbdb5096c7ee4533b234a8bf90e9271341048b42ffcd6c0e143ceR785)!